### DISCUSSION TITLE
Nested Modules, why the need of such complex implementation?

### DISCUSSION THREAD

**Discussion by madalinignisca** [area17/twill#1227] [Q&A]
Hi. I'm trying to understand the need of Kalnoy\\Nestedset for having nested entries for a Module.

From my point of view having parent_id and order already handles every scenario I can think about when I need to have some resources in a Tree.

The LFT and RGT, I can't find them any use if I have the order set right.

But never mind about it.

Problem I found with Kanoy's implementation, is that the data must be perfect from start. Let's say I need to import data from another source, that had no left and right. The package enforces me to ensure I set LFT and RGT in the import process, but that data is of no use when all elements are already in the root. Even if I make them in a tree, and already put the correct parent_id and order, the TREE is perfect to use.

Kanoy's package, fails to save a new tree if those values are missing.

Actually it has a dedicated method that simply throws an exception if any of them are missing. I would have found the correct approach to calculate them on the fly if are NULL or 0 (which is default I see), but instead of saving a TREE from fresh entries without LFT and RGT, simply fails with ...

What is the real use case for must having this complex TREE handling instead of a simple approach for parent_id + order? (I'm maintaining also a large product catalog with this simple method with millions of entries and never failed to manipulate easy the tree).

**Comment by madalinignisca** [comment]
To add a bit of more info: I imported data, and set that _left would be resource n-1 and right would be n+1 (all ids exist). Still, when I want to save, I get this error:

```
[2021-11-01 11:09:46] production.ERROR: Cannot move node into itself. {"userId":1,"exception":"[object] (LogicException(code: 0): Cannot move node into itself. at /var/www/html/vendor/kalnoy/nestedset/src/QueryBuilder.php:548)
[stacktrace]
#0 /var/www/html/vendor/kalnoy/nestedset/src/NodeTrait.php(591): Kalnoy\\Nestedset\\QueryBuilder->moveNode()
```
What is the default _lft _rgt to use when importing data from a system that had only parent_id and position (the position in current level it is)?

I think the implementation in Kanoy might cause even more unknown yet problems.

**Reply by ifox** [reply]
Hi @madalinignisca, you actually don't need to ever manually figure out those values. The package has a `fixTree` helper that will compute them out of `parent_id`: https://github.com/lazychaser/laravel-nestedset#migrating-from-basic-parentage-info. The rationale behind using this package is query performance. Read about nested sets, they are a great way to deal with hierarchical data.

**Reply by ifox** [reply]
The package also comes with a lot of powerful functionality around trees that we do not want to reinvent the wheel for.

**Reply by madalinignisca** [reply]
So, every import should be finished with a fixTree? for the API I expose, I'm doing direct DB queries, which performed like 3-5x times faster than using Eloquent/Models in all cases.
